// react-native-coverflow/lib/Coverflow.js
Object.defineProperty(exports,"__esModule",{value:true});var _slicedToArray=function(){function sliceIterator(arr,i){var _arr=[];var _n=true;var _d=false;var _e=undefined;try{for(var _i=arr[typeof Symbol==='function'?Symbol.iterator:'@@iterator'](),_s;!(_n=(_s=_i.next()).done);_n=true){_arr.push(_s.value);if(i&&_arr.length===i)break;}}catch(err){_d=true;_e=err;}finally{try{if(!_n&&_i["return"])_i["return"]();}finally{if(_d)throw _e;}}return _arr;}return function(arr,i){if(Array.isArray(arr)){return arr;}else if((typeof Symbol==='function'?Symbol.iterator:'@@iterator')in Object(arr)){return sliceIterator(arr,i);}else{throw new TypeError("Invalid attempt to destructure non-iterable instance");}};}();var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _react=require('react');var _react2=_interopRequireDefault(_react);
var _reactNative=require('react-native');
var _ViewPropTypes=require('react-native/Libraries/Components/View/ViewPropTypes');var _ViewPropTypes2=_interopRequireDefault(_ViewPropTypes);

var _constants=require('./constants');







var _Item=require('./Item');var _Item2=_interopRequireDefault(_Item);
var _clamp=require('./clamp');var _clamp2=_interopRequireDefault(_clamp);
var _fixChildrenOrder=require('./fixChildrenOrder');var _fixChildrenOrder2=_interopRequireDefault(_fixChildrenOrder);
var _convertSensitivity=require('./convertSensitivity');var _convertSensitivity2=_interopRequireDefault(_convertSensitivity);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _objectWithoutProperties(obj,keys){var target={};for(var i in obj){if(keys.indexOf(i)>=0)continue;if(!Object.prototype.hasOwnProperty.call(obj,i))continue;target[i]=obj[i];}return target;}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}

var styles=_reactNative.StyleSheet.create({
container:{
justifyContent:'center',
alignItems:'center'}});var



Coverflow=function(_Component){_inherits(Coverflow,_Component);
































function Coverflow(props){_classCallCheck(this,Coverflow);var _this=_possibleConstructorReturn(this,(Coverflow.__proto__||Object.getPrototypeOf(Coverflow)).call(this,
props));_initialiseProps.call(_this);

var sensitivity=(0,_convertSensitivity2.default)(props.sensitivity);
_this.scrollPos=props.initialSelection;
var scrollX=new _reactNative.Animated.Value(props.initialSelection);
_this.state={
width:0,
sensitivity:sensitivity,
scrollX:scrollX,
selection:props.initialSelection,
children:(0,_fixChildrenOrder2.default)(props,props.initialSelection)};return _this;

}_createClass(Coverflow,[{key:'componentWillMount',value:function componentWillMount()

{var _this2=this;var _state=
this.state,scrollX=_state.scrollX,sensitivity=_state.sensitivity;
this.scrollListener=scrollX.addListener(this.onScroll);

this.panResponder=_reactNative.PanResponder.create({
onStartShouldSetPanResponder:function onStartShouldSetPanResponder(){return true;},
onMoveShouldSetPanResponder:function onMoveShouldSetPanResponder(evt,gestureState){return(


Math.abs(gestureState.dx)>10);},

onPanResponderGrant:function onPanResponderGrant(){
scrollX.extractOffset();
},
onPanResponderTerminationRequest:function onPanResponderTerminationRequest(){return true;},
onPanResponderMove:function onPanResponderMove(evt,gestureState){
scrollX.setValue(-(gestureState.dx/sensitivity));

},
onPanResponderRelease:function onPanResponderRelease(evt,gestureState){
scrollX.flattenOffset();

var count=_react.Children.count(_this2.props.children);
var selection=Math.round(_this2.scrollPos);


if(selection>0&&selection<count-2&&Math.abs(gestureState.vx)>1){
var velocity=-Math.sign(gestureState.vx)*(
(0,_clamp2.default)(Math.abs(gestureState.vx),3,5)/sensitivity);
var deceleration=_this2.props.deceleration;

_reactNative.Animated.decay(scrollX,{
velocity:velocity,
deceleration:deceleration}).
start(function(_ref){var finished=_ref.finished;

if(finished){
_this2.snapToPosition();
}
});
}else{
_this2.snapToPosition();
}
}});

}},{key:'componentWillReceiveProps',value:function componentWillReceiveProps(

nextProps){

var sensitivity=(0,_convertSensitivity2.default)(nextProps.sensitivity);
var selection=(0,_clamp2.default)(this.state.selection,0,_react.Children.count(nextProps.children)-1);
var children=(0,_fixChildrenOrder2.default)(nextProps,selection);

if(this.state.selection!==selection){
this.state.scrollX.setValue(selection);
}

this.setState({
selection:selection,
sensitivity:sensitivity,
children:children});

}},{key:'componentWillUnmount',value:function componentWillUnmount()

{
this.state.scrollX.removeListener(this.listenerId);
}},{key:'render',value:function render()





















































































{var _props=










this.props,style=_props.style,rotation=_props.rotation,midRotation=_props.midRotation,scaleDown=_props.scaleDown,scaleFurther=_props.scaleFurther,perspective=_props.perspective,spacing=_props.spacing,wingSpan=_props.wingSpan,props=_objectWithoutProperties(_props,['style','rotation','midRotation','scaleDown','scaleFurther','perspective','spacing','wingSpan']);var
children=this.state.children;

return(
_react2.default.createElement(_reactNative.View,_extends({
style:[styles.container,style]},
props,{
onLayout:this.onLayout},
this.panResponder.panHandlers),

children.map(this.renderItem)));


}}]);return Coverflow;}(_react.Component);Coverflow.propTypes={style:_ViewPropTypes2.default.style,sensitivity:_react.PropTypes.oneOf([_constants.SENSITIVITY_LOW,_constants.SENSITIVITY_NORMAL,_constants.SENSITIVITY_HIGH]),deceleration:_react.PropTypes.oneOf([_constants.DECELERATION_NORMAL,_constants.DECELERATION_FAST]),initialSelection:_react.PropTypes.number,spacing:_react.PropTypes.number,wingSpan:_react.PropTypes.number,rotation:_react.PropTypes.number,midRotation:_react.PropTypes.number,perspective:_react.PropTypes.number,scaleDown:_react.PropTypes.number,scaleFurther:_react.PropTypes.number,children:_react.PropTypes.arrayOf(_react.PropTypes.element).isRequired,onPress:_react.PropTypes.func,onChange:_react.PropTypes.func.isRequired};Coverflow.defaultProps={initialSelection:0,style:undefined,sensitivity:_constants.SENSITIVITY_NORMAL,deceleration:_constants.DECELERATION_NORMAL,spacing:100,wingSpan:80,rotation:50,midRotation:50,perspective:800,scaleDown:0.8,scaleFurther:0.75,onPress:undefined};var _initialiseProps=function _initialiseProps(){var _this3=this;this.onScroll=function(_ref2){var value=_ref2.value;_this3.scrollPos=value;var count=_this3.state.children.length;var newSelection=(0,_clamp2.default)(Math.round(value),0,count-1);if(newSelection!==_this3.state.selection){_this3.setState({selection:newSelection,children:(0,_fixChildrenOrder2.default)(_this3.props,newSelection)});}};this.onLayout=function(_ref3){var nativeEvent=_ref3.nativeEvent;_this3.setState({width:nativeEvent.layout.width});};this.onSelect=function(idx){if(idx===Math.round(_this3.scrollPos)){if(_this3.props.onPress){_this3.props.onPress(idx);}}else{_this3.snapToPosition(idx);}};this.snapToPosition=function(){var pos=arguments.length>0&&arguments[0]!==undefined?arguments[0]:_this3.scrollPos;var _state2=_this3.state,scrollX=_state2.scrollX,children=_state2.children;var count=children.length;var finalPos=(0,_clamp2.default)(Math.round(pos),0,count-1);if(finalPos!==_this3.scrollPos){_this3.props.onChange(finalPos);_reactNative.Animated.spring(scrollX,{toValue:finalPos}).start();}};this.renderItem=function(_ref4){var _ref5=_slicedToArray(_ref4,2),position=_ref5[0],item=_ref5[1];if(!_this3.state.width){return null;}var scrollX=_this3.state.scrollX;var _props2=_this3.props,rotation=_props2.rotation,midRotation=_props2.midRotation,perspective=_props2.perspective,children=_props2.children,scaleDown=_props2.scaleDown,scaleFurther=_props2.scaleFurther,spacing=_props2.spacing,wingSpan=_props2.wingSpan;var count=_react.Children.count(children);return _react2.default.createElement(_Item2.default,{key:item.key,scroll:scrollX,position:position,count:count,spacing:spacing,wingSpan:wingSpan,rotation:rotation,midRotation:midRotation,perspective:perspective,scaleDown:scaleDown,scaleFurther:scaleFurther,onSelect:_this3.onSelect},item);};};exports.default=


Coverflow;
